### PR TITLE
enable dirty regions for full screen video

### PR DESCRIPTION
**Pull request by FernetMenta**
Moves video rendering (fullscreen) from application to GUIWindowFullScreen. By doing this dirty regions work on fullscreen mode and the platform has separate layers for video and ui.

@popcornmix iirc you still need some options to limit fps of gui, right?

### CODE REVIEW

**Comment by fritsch** [comment]
When watching LiveTV in preview window - there is no picture at all.

**Comment by FernetMenta** [comment]
@fritsch VideoControl is fixed now.

**Comment by fritsch** [comment]
Yeah, I expected a one liner - works as expected. Btw. did you check with intel_gpu_top or something that we get measurable results? E.g. reduced gpu usage? More throughput?

Btw. @smallint wants to contact you for some specialities the cubox (and most likely other hw) will need, having a high performance fb on fb1 and another one fb0 which needs quite heavy operations to swap and sync.

**Comment by fritsch** [comment]
Wait, something else.

The first channel after a "stop" does not show a picture. When switching to fullscreen and back, picture is there. Subsequent channels work if you don't press stop between them.

**Comment by FernetMenta** [comment]
fixed again.
This change has no performance impact on systems with only a single buffer for rendering. Systems like Pi or Cubux will benefit from it.

**Comment by smallint** [comment]
I checked this PR and it seems to work regarding eglSwapBuffers which are not called anymore in fullscreen if nothing is visible. I incorporated the change needed in CLinuxRendererGLES::HasFrame() as discussed with @FernetMenta.

But if I press enter and the OSD should show up only the currenly updated controls are visible. Same is true for 'o'. Since the renderer calls glClear all controls are removed but not the onces currently dirty. Furthermore m_rotatedDestCoords in RenderIMXMAPTexture are somehow wrong? In LiveTV I get now values like "0 1082726400 0 1080074240" for "x0 y0 x1 y1" if displayed not fullscreen. In fullscreen I get always "0 0 0 0".

**Comment by FernetMenta** [comment]
what dirty regions algo did you choose? in general we render all controls if there is only a single one dirty because there is no preserve back buffer.

**Comment by smallint** [comment]
```
 <gui>
    <algorithmdirtyregions>1</algorithmdirtyregions>
    <nofliptimeout>1000</nofliptimeout>
  </gui>
```

**Comment by smallint** [comment]
If I open the OSD and move with left and right through the buttons I can see only the currently updated buttons but all others are cleared.

**Comment by fritsch** [comment]
@Fernetmenta: The last squash patch needs #ifdef HAS_VIDEO_PLAYBACK

2015-01-02 13:10 GMT+01:00 smallint <notifications@github.com>:

> If I open the OSD and move with left and right through the buttons I can
> see only the currently updated buttons but all others are cleared.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6090#issuecomment-68522745>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by FernetMenta** [comment]
@smallint algo is wrong. as you observe, your system has no back buffer preserve. hence you need to invalidate all controls.

@fritsch I think i will remove this outdated HAS_VIDEO_PLAYBACK in the near future. complete nonsense in this context here.

**Comment by fritsch** [comment]
@FernetMenta Might be, but it does not link without that :-) I did not care about the semantics.

**Comment by FernetMenta** [comment]
it links here.

**Comment by fritsch** [comment]
Okay, try make clean and full make again, then you might see.

2015-01-02 13:27 GMT+01:00 Rainer Hochecker <notifications@github.com>:

> it links here.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6090#issuecomment-68523462>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

**Comment by smallint** [comment]
@FernetMenta Actually the system has back buffer preserve. I checked it with another application and also the log does not tell otherwise. The renderer always calls glClear, doesn't it? And that destroys the back buffer anyhow.

I tried with "3" and it works that all controls are rendered if invalidated. But the next frame will remove them again completely (glClear?).

**Comment by smallint** [comment]
I could fix it for me with
```
if (m_bRenderBypass)
{
  ...
  g_renderManager.Render(m_bHasRendered, 0, 255);
  ...
}
```

Then it is working with both algos.

**Comment by popcornmix** [comment]
@FernetMenta We do enable back buffer preserve (https://github.com/xbmc/xbmc/pull/3021) and the Pi certainly respects that flag.
I'll check if algorithmdirtyregions=1 works with video with this PR (it used to leave artefacts when video had black bars).

**Comment by Memphiz** [comment]
seems to work normal for osx and ios (just tested the osd controls and dialogs ...)

**Comment by popcornmix** [comment]
@smallint where were you adding that code?

**Comment by FernetMenta** [comment]
@popcornmix CGUIWindowFullScreen::FrameMove()
@smallint thanks for the fix.

**Comment by smallint** [comment]
@popcornmix In GUIWindowFullScreen.cpp:555

**Comment by smallint** [comment]
Now playback with and without GUI works very well and with almost full performance (no GUI). Next small issue is overlay rendering such as teletext. It only updates were OSD controls are dirty.

**Comment by popcornmix** [comment]
@smallint do videos with black bars clear background correctly?
I find UI shows in black bars with algorithmdirtyregions=1 on Pi.
Also I get artefacts on other screens like system info with algorithmdirtyregions=1 (looks to be an issue with alpha being applied more than once).

**Comment by smallint** [comment]
@popcornmix Yes. E.g. videos with 4:3 ratio do show correctly. I must admit that I am clearing the background manually if the cropping rect changes. With "background" I refer to another framebuffer device than the GL context uses. It is probably similar to the Pi hardware layering. In the GL context I just cut a hole with alpha=0 into the paint area if not fullscreen.

**Comment by FernetMenta** [comment]
@smallint does the teletext dialog set its dirty flag? Maybe you need to override the DoProcess method and call MarkDirtyRegion when something has changed.

**Comment by smallint** [comment]
Where is the teletext dialog implemented? I will check that.

**Comment by FernetMenta** [comment]
https://github.com/xbmc/xbmc/blob/master/xbmc/video/dialogs/GUIDialogTeletext.h

**Comment by smallint** [comment]
Found it. Indeed, it does not set the dirty flag. Doing that helps but "esc" for hiding it has then a very high latency then. I need to press esc twice to make it disappear after a couple of seconds.

**Comment by FernetMenta** [comment]
how did you set it? dependent on m_TextDecoder.NeedRendering()?

**Comment by smallint** [comment]
Yes, but I forgot to check m_bClose in DoProcess. Now it is working.

**Comment by FernetMenta** [comment]
can you pr this change to this branch?

**Comment by smallint** [comment]
would do that but how can I fork your repo again if I have already the xbmc repo forked? or is there another way to issue a PR into your branch?

**Comment by FernetMenta** [comment]
no need to fork my repo. you can select the target when creating a pr. if you have trouble with that, just push to your branch and I'll cherry-pick

**Comment by smallint** [comment]
OK, done

**Comment by FernetMenta** [comment]
updated with the work of @smallint 
@popcornmix are you ok with this?

**Comment by smallint** [comment]
There is still the issue of the codecinfo overlay I am currently investigating. I think the GUIControl does not update the hidden state correctly ...

**Comment by smallint** [comment]
I applied the following fix which helps to clear the background correctly but honestly I do not understand why it helps ;)

```
diff --git a/xbmc/guilib/GUIControl.cpp b/xbmc/guilib/GUIControl.cpp
index e3d04db..bfd203e 100644
--- a/xbmc/guilib/GUIControl.cpp
+++ b/xbmc/guilib/GUIControl.cpp
@@ -511,6 +511,7 @@ void CGUIControl::SetVisible(bool bVisible, bool setVisState)
   {
     m_forceHidden = !bVisible;
     SetInvalid();
+    if (m_forceHidden) MarkDirtyRegion();
   }
   if (m_forceHidden)
   { // reset any visible animations that are in process
```

**Comment by smallint** [comment]
Beside the open issues I like the changes of this PR which help a lot and work very well. Thanks!

**Comment by FernetMenta** [comment]
what exactly is the issue with the codec info?

**Comment by smallint** [comment]
it does not hide completely. either label_row1 stays or parts of the transparent image. the above patch fixes it but I think there must be a better approach. that only happens if glClear is not called in the renderer with the new fullscreen path.

**Comment by topfs2** [comment]
Jumping in a bit late here but regarding not preserving back buffer this is
why we track 3 changes back. So that we know what changed in a triple
buffer setting. If it completely wipe the back buffer under your nose its
impossible though, I haven't seen a system which does that though?
Den 2 jan 2015 22:12 skrev "smallint" <notifications@github.com>:

**Comment by FernetMenta** [comment]
@smallint I reworked this, could you try again? You may need to adapt LinuxRendererGLES. There is an additional path for methods rendering to a separate video layer.

**Comment by smallint** [comment]
@FernetMenta I tried it and it is even worse. I needed to implement RenderUpdateVideo. The player window (not fullscreen) does not work anymore as RenderUpdateVideo is not called? In fullscreen the controls are all erased due to the glClear that is done all the time now (scissor). Actually the clear should be done in the window itself that wants to render content. The renderer should only render its image since it does not now if its window needs a clear or not. Whenever the GUI changes, the player window (fullscreen or not) should clear its content depending on IsGuiLayer(), shouldn't it? Currently your are wiping the back buffer at every frame. Did I do something wrong with that implementation?

**Comment by smallint** [comment]
I was able to make fullscreen work by implementing

```
void CGUIWindowFullScreen::ClearBackground()
{
  if (!g_renderManager.IsGuiLayer())
    g_graphicsContext.Clear(0);
}
```

The renderer does not need to clear its background anymore. It is anyhow too late in AfterRender to clear the background since all controls were rendered already.

**Comment by smallint** [comment]
The following patch is required for GUIVideoControl.cpp
```
--- a/xbmc/guilib/GUIVideoControl.cpp
+++ b/xbmc/guilib/GUIVideoControl.cpp
@@ -69,7 +69,7 @@ void CGUIVideoControl::Render()
 
 #ifdef HAS_VIDEO_PLAYBACK
     color_t alpha = g_graphicsContext.MergeAlpha(0xFF000000) >> 24;
-    g_renderManager.Render(false, 0, alpha);
+    g_renderManager.Render(false, 0, alpha, g_renderManager.IsGuiLayer());
 #else
     ((CDummyVideoPlayer *)g_application.m_pPlayer->GetInternal())->Render();
 #endif
```
I also tried to play with MarkDirtyRegion to clear the background only when required but that did not work. We could easily speed up windowed playback as well. When setting MarkDirtyRegion() only if IsGuiLayer() and call Clear(0) along with a correct scissor rect inside Render it destroys the window contents, e.g. the main screen when the player clock display changes every second. I don't know if the GUI rendering uses the Z buffer or if controls are rendered in Z order. I think there are still plenty optimizations in terms of rendering possible once one understands the entire rendering system.

**Comment by FernetMenta** [comment]
@smallint GUIWindowFullScreen is not supposed to clear its background. It sets m_clearBackground to 0. Why do you need this modification?

```
void CGUIWindowFullScreen::ClearBackground()
{
  if (!g_renderManager.IsGuiLayer())
    g_graphicsContext.Clear(0);
}
```

**Comment by FernetMenta** [comment]
> Whenever the GUI changes, the player window (fullscreen or not) should clear its content depending on IsGuiLayer(), shouldn't it?

Not for those renderers which render to the gui layer. The background should be cleared for those rendering to a separate video layer.

**Comment by smallint** [comment]
@FernetMenta Because the background needs to be cleared *before* the rendering of child controls otherwise the clear in RenderUpdateVideo destroys the control rendering.

> Not for those renderers which render to the gui layer. The background should be cleared for those rendering to a separate video layer.

Exactly why I had to add the ClearBackground method.

**Comment by smallint** [comment]
That brings up another question: CGUIWindowFullScreen does now render the non gui path in AfterRender. I would like to test the same for CGUIVideoControl but that is a control and does not expose that method. I managed to clear the background in that control correctly and would like to render the video which does not happen currently because Render is not called anymore (which is fine).

**Comment by FernetMenta** [comment]
@smallint the GuiVideoControl is supposed to render to gui layer. A skin can make use of it multiple times in a window.

**Comment by smallint** [comment]
@FernetMenta OK, I see. What happens if a skin instantiates the control twice in a view? FrameFinish would be called twice as well, wouldn't it?

As a test I hacked Render(false, 0, 255, false) and FrameFinish into the application class after g_windowManager.AfterRender() and removed it from CGUIWindowFullScreen. Now the playback even in the main screen renders very smoothly and eglSwapBuffers is only called once a second or the clock. Would be nice to find an official way to incorporate such optimizations.

**Comment by FernetMenta** [comment]
you should post your changes somewhere or comment on this code. It is hard to follow what exactly you did.

**Comment by smallint** [comment]
Its here: https://github.com/smallint/xbmc/commits/perf-test-fm. This is a mockup for the new IMX6 rendering, not finalized.

**Comment by FernetMenta** [comment]
@smallint I squashed the changes and made the diff look nicer by correcting added/removed spaces and new lines. Since you have done more work I made you the author. Could you please test if it still works.

**Comment by smallint** [comment]
@FernetMenta I built a few minutes ago an Arch package and run it currently on my operational box. I will furthermore pull your branch again, diff my changes and test this build. Thanks for the effort put into this PR!

**Comment by smallint** [comment]
https://github.com/smallint/xbmc/commit/31a8af4f1407c0c0362eba4059fc1ff58fb09e51 is still required to hide e.g. the codecinfo overlay correctly.

**Comment by FernetMenta** [comment]
updated
@popcornmix are you fine with this?
jenkins build this please

**Comment by popcornmix** [comment]
No objections in principle. I've tested earlier versions and didn't spot any regressions on Pi.
algorithmdirtyregions=1 wasn't usable, but this PR doesn't make it any worse.
Let me have a play with latest version...

**Comment by smallint** [comment]
@popcornmix Does the Pi render video with bypass? If so, I would be interested if you spot any changes in performance in particular when you navigate to the main screen with playback in the background.

**Comment by FernetMenta** [comment]
@popcornmix I tried with dirty regions algo 1 on my Pi and noticed no issues. What exactly did not work for you?

**Comment by popcornmix** [comment]
Pi uses bypass when omxplayer acceleration is enabled, and uses the MMAL renderer (which also renders to a separate layer) when disabled.

**Comment by popcornmix** [comment]
Did you just change algorithmdirtyregions=1 or did you change nofliptimeout or other settings?

I found that alpha seemed to "build up". Bits that were redrawn were okay, but some non-redrawn parts with alpha seemed to get thicker around the edges. The system info screen was particularly bad with some flickering rectangles.

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp [hunk 1]
```diff
@@ -3046,5 +3050,13 @@ void CLinuxRendererGLES::AddProcessor(CDVDVideoCodecIMXBuffer *buffer, int index
 }
 #endif
 
+bool CLinuxRendererGLES::IsGuiLayer()
+{
+  if (m_format == RENDER_FMT_BYPASS)
```
Comment (FernetMenta) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: @smallint why is  RENDER_FMT_IMXMAP missing here?

**Comment by FernetMenta** [comment]
I changed only algorithmdirtyregions=1 and used oxmplayer. MMAL is not merged to mainline yet, right?

**Comment by popcornmix** [comment]
Yes MMAL is in mainline. Just disable omxplayer in acceleration settings and it will use MMAL for decode and render. Disable MMAL acceleration and it will just use MMAL for render of YUV.

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp (hunk 1 above)
Comment (smallint) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: @FernetMenta because that is part of the new rendering and neither upstream nor in a PR yet. it is just my branch that requires "bypass".

**Comment by FernetMenta** [comment]
@popcornmix I guess MMAL renderer should override IsGuiLayer(), right? Did you try this?

File: xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp]: ahh, understood

**Comment by popcornmix** [comment]
https://github.com/popcornmix/xbmc/commit/aeedaa41078834c2ee9f08abcf347695fd49c4a2
means I don't see GUI (e.g. codec info) when video is playing.
I see GUI without that.

**Comment by FernetMenta** [comment]
@popcornmix I think you also need to remove the clear: https://github.com/popcornmix/xbmc/blob/aeedaa41078834c2ee9f08abcf347695fd49c4a2/xbmc/cores/VideoRenderers/MMALRenderer.cpp#L361

clear is done by the window when rendering is started.

**Comment by smallint** [comment]
... and whenever anything changes. The GUI now takes the responsibility to clear the background for you for the !IsGuiLayer() path. Otherwise you will call clear in the render code which happens after the GUI is rendered -> GUI wiped.

**Comment by popcornmix** [comment]
Removed glClear and codec overlay is fixed.
Whenever subs appear I lose the video (presumably overlaid with non-alpha black)
Video reappears whenever subs are no showing.

**Comment by smallint** [comment]
I can confirm that. When subs are enabled the GUI is not cleared anymore.

**Comment by FernetMenta** [comment]
could you try again please? I also added the MMAL changes

**Comment by FernetMenta** [comment]
@koying could you test on Android if this breaks something?

**Comment by popcornmix** [comment]
I'm not seeing any functional regressions with this version. It will be in tonight's newclock4 test build.

I think I'm seeing a slight CPU % increase with this PR in place when playing videos (with default algorithmdirtyregions=3). This is just watching top while playing, so not hugely accurate, so I'd like to find a more reliable measure. It may just be noise.

**Comment by FernetMenta** [comment]
do you observe this with omx or dvdplayer?

**Comment by popcornmix** [comment]
A bit with both. Without PR I was measuring, from top:
40% idle with omxplayer, 17% idle with dvdplayer
After PR 
37% idle with omxplayer, 15% idle with dvdplayer
Now, reading top isn't particularly stable so I'd like to find a more repeatable metric.

**Comment by popcornmix** [comment]
I've been testing more carefully. I'm playing a test file (prometheus dream sequence) and wind down the arm cpu frequency until we get a failure (omxplayer pausing due to underrun, or dvdplayer setting drop state).

omxplayer does seem unchanged by this PR. Consistently got 535MHz before and after.
dvdplayer does show a regression. I used an easier file and before this PR it passed at 505MHz. After this PR it passes at 595MHz.

So I think there is something which harms dvdplayer perfromance on Pi in here.

**Comment by popcornmix** [comment]
A bit more testing. I'm using this patch to limit gui update rate when playing video:
https://github.com/popcornmix/xbmc/commit/65e3fae4d4eade44c06ed4224102976ae0cd58ec

Before PR, with omxplayer with an overlay ('i' key info) visible, the limit is 565MHz when gui update limit is disabled. With gui limited to 5fps this improves to 545MHz.

After PR, with gui limit disabled I get 580MHz, with the limit in place we get 545MHz.

So I think there is a performance hit from this PR when using omxplayer with an overlay visible.

**Comment by popcornmix** [comment]
With MMAL, no visible gui overlay:
Before PR: unlimited gui:  520MHz
Before PR: 10Hz gui:  500MHz
After PR, unlimited gui: 535MHz
After PR, 10Hz gui: 555MHz

This PR + gui limiting hack causes a big performance hit.
Without the gui limiting hack, there is still a smaller performance hit.

**Comment by smallint** [comment]
This PR helps a lot for rendering my path but I must admit that CPU usage is not a big concern in my particular case. Nonetheless the Pi performance hit needs to be sorted out!

When I open the codecinfo overlay I see sometimes frame rates up to 100 fps (at least higher than 50fps) which is impossible since my display is running at 50 hz and the video playback vsyncs. So bypassing the GL buffer swap (g_Windowing.PresentRender) seems to introduce a higher CPU usage or just confuses the fps updater. I still haven't tested the latest changes though ...

**Comment by FernetMenta** [comment]
@popcornmix this 10Hz gui hack is wrong now. renderManager.FrameMove is called by application::FrameMove on every cycle. Now you call it twice.
Another thing I was wondering is why you send the video buffer in FlipPage (MMAL) and not in RenderUpdate. Feels wrong somehow. FlipPage only should advance the buffer.

**Comment by smallint** [comment]
@FernetMenta You remark would affect my implementation, too. Where do you recommend should I do the vsync'd page flip (framebuffer)? Currently I am doing it in CLinuxRendererGLES::FlipPage which reads like the correct function to do that. Should I move it to CLinuxRendererGLES::RenderIMXMAPTexture which is called from RenderUpdateVideo?

**Comment by FernetMenta** [comment]
@smallint yes move this function. Have a look at CLinuxRendererGLES::FlipPage. It does nothing but advance the buffer. (was not me who has chosen this misleading name)

@popcornmix could you try with this reverted: https://github.com/FernetMenta/xbmc/commit/b8ef805c45a4d83ec297393685eaac549afb2689
The current code holds the gfx lock while waiting in WaitPresentTime. This can be a long time, hence I consider this no-go. Maybe you observed less CPU load just because other threads were blocked.

**Comment by smallint** [comment]
> Have a look at CLinuxRendererGLES::FlipPage. It does nothing but advance the buffer.

True, but that depends on eglSwapBuffers which is managed by the application. Frame buffer swaps outside EGL need to be handled differently and here I am looking for the best place to sync with the current player that it displays the right buffer and the right time. I am going to move the function and test it. I found the FlipPage approach a while ago in https://github.com/rabeeh/xbmc/blob/master/xbmc/cores/VideoRenderers/DoveOverlayRenderer.cpp#L276 and have taken it as best practice ...

**Comment by FernetMenta** [comment]
Note that presentation time is calculated for double buffering. When WaitPresentTime returns, the frame rendered in the previous cycle (not the one just rendered) in on screen. The actual swapping of front/back buffers happens while we wait in WaitPresentTime (when running with videorefclock).
We don't want to block in egl/glx swapBuffers. Unfortunately systems without a video reference clock implementation behave slightly different because we don't know when the actual swap occurs.

**Comment by smallint** [comment]
@FernetMenta I can confirm that your latest changes do work with respect to subs. Also thanks for pointing out how the rendering works.

**Comment by popcornmix** [comment]
@FernetMenta Removing the FrameMove call from the gui limiter avoids that bad behaviour.
My original benchmark counted drops to detect a failure. But we always get a handful at the start of playback which makes it a bit hard to judge when it's good/bad. 
I'm now counting skips which seem more stable. We always get 1 at start, so I'm treating a second as a fail.

Now I get (with MMAL and no overlay):
Before PR:
Unlimited 620MHz
10 Hz      530MHz

After PR:
Unlimited 580MHz
10 Hz      530MHz

So this benchmark looks better. The PR seems to be helping the case when the gui limiting hack is not present, and doesn't harm the case when it is present. I'd like to recheck omxplayer and test with an overlay up.

Removing the "exit gfx lock" commit didn't obviously help.

**Comment by popcornmix** [comment]
Same test as before but bringing up overlay ('i' key).
Before PR:
Unlimited 890MHz
10 Hz 700MHz

After PR:
Unlimited 940MHz
10 Hz 870MHz

If I revert "exit gfx lock" it does help:
Unlimited 860MHz
10 Hz 830MHz

but we're still significantly below gui-limited performance from before this PR. This is with FrameMove removed from gui limiter.

Is it possible rendering from RenderUpdate rather than FlipPage will help?

**Comment by FernetMenta** [comment]
> Is it possible rendering from RenderUpdate rather than FlipPage will help?

not sure if it has any impact on performance but should be changed anyway.

> If I revert "exit gfx lock" it does help

Could you investigate further? I would guess something like GuiInfoManager puts the load on the CPU if not blocked.

**Comment by smallint** [comment]
After testing a couple of days I only found a small issue that seems to be related. When I start LiveTV (not fullscreen) the destination render rect is sometimes not resized correctly. It results in a right padding of about 10-20% of the video control. Switching to fullscreen and back fixes that. I can almost exclude an iMX specific render bug as I am pulling the RenderManager rects at each frame blit. Probably some internal layout changes are not finished yet while the VideoControl already stopped marking itself as dirty. This is really a minor issue and apart from that it is working great.

**Comment by smallint** [comment]
A nice addition would be if the VideoRenderer could be informed about changes of the flip state. This could be useful for further optimizations. E.g.
```
g_renderManager.UpdateGUIFlipState(flip);
```
in Application.cpp. That calls a m_pRenderer method which in turn implements the logic.

**Comment by FernetMenta** [comment]
I want to remove all rendermanager calls from application and remove the global g_renderManager. We can implement a callback in windowing if flip state is desired

**Comment by smallint** [comment]
OK. Then just one remark if g_renderManager goes away: currently I am calling
```
g_renderManager.GetVideoRect(srcRect, destRect)
```
in the codec. May sound weird but to queue as much processed frames as possible blitting is done in the codec. Then I need to find a way to access that information. Will you do it as part of this PR or in another one once this is merged?

**Comment by FernetMenta** [comment]
That is work for another pr. Could you provide more information on this requirement?

**Comment by smallint** [comment]
Sure, should I comment here since it is not related to this PR?

**Comment by FernetMenta** [comment]
Better in the forum

**Comment by FernetMenta** [comment]
@popcornmix do you agree that we merge this and then investigate and fix the remaining issues on Pi? This is a big improvement for iMX and I don't think there is anything wrong with this pr.

**Comment by popcornmix** [comment]
I do worry that I won't be able to fix the performance regression and the Pi builds will have to maintain a revert of this PR indefinitely.

You suggested investigating GuiInfoManager. Can you provide any more information on what I should be logging?

**Comment by FernetMenta** [comment]
First I would not call this really a regression because
1) the gui hack is not in master
2) the root cause is certainly somewhere else

You confirmed that leaving the gfx lock causes an increase of CPU load. Hence the current performance of the Pi is based on faulty behavior of the application. This has to be fixed. What thread puts more load on the CPU? My assumption is that it might be GuiInfoManager because it is known to be a beast with its polling of everything.

**Comment by smallint** [comment]
Just a guess but since windowing flipping is optimized, https://github.com/xbmc/xbmc/blob/master/xbmc/rendering/gles/RenderSystemGLES.cpp#L264 is not called anymore which does some sleeps to sync the framerate. Might that cause higher CPU load? If so then Application::flip = true if Pi is used could fix it.

**Comment by FernetMenta** [comment]
the main thread either should sleep in RenderManager::FrameFinish or FrameWait:
https://github.com/xbmc/xbmc/blob/master/xbmc/Application.cpp#L2230

**Comment by smallint** [comment]
Just for my understanding: RenderManager itself does not wait actively, does it? If your render path would not wait for vsync then the render loop would be quite CPU intensive. I had the case for 25fps streams that RenderUpdateVideo was called at a rate of 50fps (because my path blocked for vsync). So every frame was processed and blitted twice. If I would not block for vsync then it would be called more often unless someone else is throttling the loop (Application?).

**Comment by FernetMenta** [comment]
The render thread waits in calls to rendermanager. The loop is executed once for each progressive frame and twice for some deinterlacing methods. Can you check if you had deint flag set?

**Comment by smallint** [comment]
I haven't had the deint flag set, for sure. At that time I didn't even know that it existed ;)
I can do more tests later and post some debug output later.

**Comment by FernetMenta** [comment]
as discussed with @popcornmix offline. we get this in now and I will help him investigating the issue on Pi once he has time to work on this.

jenkins build this please

**Comment by smallint** [comment]
> I can do more tests later and post some debug output later.

If RenderUpdateVideo does not block at all it looks like this: http://xbmclogs.com/show.php?id=398970, whereas "r->r:" is the time between two RenderUpdateVideo calls. If it blocks for some milliseconds the situation changes as follows: http://xbmclogs.com/show.php?id=398973 whereas "r:" is the time spent in RenderUpdateVideo. I am wondering about the additional calls between the 40ms.

**Comment by FernetMenta** [comment]
not sure what your concerns are. I don't see anything unusual in this logs. In general all this rendering is async. in average you measure 40ms but it can also be 60ms follow less than 40 without having to drop a frame.

**Comment by smallint** [comment]
There is no real "concern", it is just that RenderUpdateVideo is not called every 40ms (25fps) but more often. I can't explain that better and I don't want to talk about my "feelings" ;) If I open the GUI I get more or less a call every 20ms but after closing the GUI it never recovers to every 40ms but 40-1-40-1-40-1 or similar, does that look OK? I would have expected 40-40-40-40. I can work around that, no problem, just wanted to help to find a possible issue that could affect @popcornmix.

**Comment by FernetMenta** [comment]
where exactly do you read this from? please point me to the lines in the log

**Comment by smallint** [comment]
Not from this logs. I just read it previously. Here it another log where the first number after "r:" is the buffer index: http://xbmclogs.com/show.php?id=399113.

**Comment by smallint** [comment]
I captured now two cases and logged the RenderManager states as well: http://xbmclogs.com/show.php?id=399132. Just for completeness. It seems to only happen if FrameWait is triggered.

**Comment by FernetMenta** [comment]
what exactly do you mean by "if FrameWait is triggered" ?

**Comment by smallint** [comment]
I mean the event wait:

```
bool CXBMCRenderManager::FrameWait(int ms)
{
  XbmcThreads::EndTime timeout(ms);
  CSingleLock lock(m_presentlock);
  while(m_presentstep == PRESENT_IDLE && !timeout.IsTimePast())
  {
    printf("FrameWait(%d)\n", timeout.MillisLeft());
    m_presentevent.wait(lock, timeout.MillisLeft());
  }
  return m_presentstep != PRESENT_IDLE;
}
```

**Comment by FernetMenta** [comment]
the sequence is as follows:

- FrameWait
- exit wait when a new frame is signaled by player
- FrameMove
- PrepareNextRender, this increments to the next buffer if: 
   m_Queue[idx].timestamp <= clocktime + MAXPRESENTDELAY)

so if you see the same buffer rendered twice this condition is not true

could you verify this?

**Comment by smallint** [comment]
I added an output if !next and this output is not there. Actually flip is only triggered if next is true and the previous log has flips.

```
-- idle
FrameWait(100)
-- ready
Render(gui=false)
r->r: 34
r: 0  5
-- flip
-- frame
Render(gui=false)
r->r: 6
r: 2  0
-- idle
FrameWait(100)
-- ready
Render(gui=false)
r->r: 41
r: 2  0
-- flip
-- frame
Render(gui=false)
r->r: 1
r: 1  0
-- idle
FrameWait(100)
-- ready
Render(gui=false)
r->r: 35
r: 1  2
```

Is that Render call between READY and FLIP normal?

**Comment by FernetMenta** [comment]
it is hard for me to follow because I don't see where you placed the logs. it should only render the same buffer twice if the timestamp of the next frame is too far in the future. 

> Is that Render call between READY and FLIP normal?

not sure what you mean. those are meant to be discrete states which have no "between". 
assume we are in the idle state, player delivers a new frame -> we go into ready, FrameMove calls PrepareNextRender which can bring us into FLIP, then PRESENT_FRAME, we render in this state, after render we go back to READY (if there is still a frame in the queue), or IDLE if queue is empty,

**Comment by smallint** [comment]
OK, I agree that it is hard to follow. I placed the logs where the new state is assigned. I see now that only RenderEx calls g_renderManager.Render() and will try to track that path with more logging. I come back to you when I found something.

**Comment by FernetMenta** [comment]
I think I know the reason. The sequence is wrong. After FrameWait we would need FrameMove to switch to the next buffer.

**Comment by FernetMenta** [comment]
@smallint http://pastebin.com/G3htBf99

**Comment by smallint** [comment]
Need to do more tests but yeah, looks good. You are the man!

**Comment by smallint** [comment]
I could not reproduce the wrong sequence anymore: so we found a bug? ;) Thanks a lot!

**Comment by FernetMenta** [comment]
thanks to you! hopefully this fixes the issue on the Pi too.

**Comment by smallint** [comment]
I add this patch to my tree and will credit you for it of course. Once I rebase onto baseline it will be removed.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
Build error on win unrelated
